Spigot-Server-Patches/0233-Anti-Xray.patch

```diff
@@ -0,0 +1,1383 @@
++        return true;
++    }
++
++    public PacketPlayOutMapChunkInfo getPacketPlayOutMapChunkInfo(PacketPlayOutMapChunk packetPlayOutMapChunk, Chunk chunk, int chunkSectionSelector) {
++        return null;
++    }
++
++    public void modifyBlocks(PacketPlayOutMapChunk packetPlayOutMapChunk, PacketPlayOutMapChunkInfo packetPlayOutMapChunkInfo) {
++        packetPlayOutMapChunk.setReady(true);
++    }
++
++    public void updateNearbyBlocks(World world, BlockPosition blockPosition) {
++
++    }
++}
+diff --git a/src/main/java/com/destroystokyo/paper/antixray/ChunkPacketBlockControllerAntiXray.java b/src/main/java/com/destroystokyo/paper/antixray/ChunkPacketBlockControllerAntiXray.java
+new file mode 100644
+index 00000000..e3e40efa
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/antixray/ChunkPacketBlockControllerAntiXray.java
+@@ -0,0 +1,518 @@
++package com.destroystokyo.paper.antixray;
++
++import java.util.HashSet;
++import java.util.Set;
++import java.util.concurrent.Executors;
++import java.util.concurrent.ExecutorService;
++
++import com.destroystokyo.paper.PaperWorldConfig;
++
++import net.minecraft.server.Block;
++import net.minecraft.server.BlockPosition;
++import net.minecraft.server.Blocks;
++import net.minecraft.server.Chunk;
++import net.minecraft.server.ChunkSection;
++import net.minecraft.server.DataPalette;
++import net.minecraft.server.DataPaletteGlobal;
++import net.minecraft.server.IBlockData;
++import net.minecraft.server.PacketPlayOutMapChunk;
++import net.minecraft.server.World;
++
++public class ChunkPacketBlockControllerAntiXray extends ChunkPacketBlockController {
++
++    private static ExecutorService executorServiceInstance = null;
++    private final ExecutorService executorService;
++    private final boolean asynchronous;
++    private final int engineMode;
++    private final int chunkEdgeMode;
++    private final int maxChunkSectionIndex;
++    private final IBlockData[] predefinedBlockData;
++    private final IBlockData[] predefinedBlockDataStone;
++    private final IBlockData[] predefinedBlockDataNetherrack;
++    private final IBlockData[] predefinedBlockDataEndStone;
++    private final boolean[] solidGlobal = new boolean[Block.REGISTRY_ID.a()];
++    private final boolean[] obfuscateGlobal = new boolean[Block.REGISTRY_ID.a()];
++    private final ChunkSection[] nearbyNullChunkSections = new ChunkSection[4];
++
++    public ChunkPacketBlockControllerAntiXray(PaperWorldConfig paperWorldConfig) {
++        asynchronous = paperWorldConfig.asynchronous;
++        engineMode = paperWorldConfig.engineMode;
++        chunkEdgeMode = paperWorldConfig.chunkEdgeMode;
++        maxChunkSectionIndex = paperWorldConfig.maxChunkSectionIndex;
++
++        if (asynchronous) {
++            executorService = getExecutorServiceInstance();
++        } else {
++            executorService = null;
++        }
++
++        if (engineMode == 1) {
```

Comment:
Might be worth extracting the engine modes to named constants to make things a bit clearer.